### PR TITLE
Fix DATA type parsing and add explicit bond types

### PR DESCRIPTION
**Pull request by khuston**
Regarding #763 

Changes made in this Pull Request:

**NOTE:** `atoms.bonds` is broken for atom groups without explicit bond typing
in this commit.

In order to write a LAMMPS.DATAWriter, I first need explicit typing of bonds,
angles, dihedrals, and impropers beyond the tupling of atoms types done now.
To hack bond types into the topology, I branched issue-363 and tried the
following:

1. Add optional `types` keyword to `topologyattrs.Bonds`. This is stored
in the class instance a `self._bondtypes`. A topology parser can pass `types`
to the Bonds topologyattr.
2. When `topologyattrs.Bonds` is called upon to generate a TopologyGroup,
these values are passed on via a new keyword in `TopologyGroup.__init__()`.
They are stored in the TopologyGroup instance as `self._bondtypes`.
3. I modified `TopologyGroup.__getitem__` to yield a Bond with attribute
`_bondtype`. If `_bondtype is not None` then it overrides the default behavior
of `bond.type`.
4. I modified `TopologyGroup()` calls wherever I could find them to pass on
`_bondtype` in the instantiation.

I found that the DATAParser in branch #363 is slightly broken, because
the atom types are saved as integers which breaks `select_atoms` (I think it
expects strings, and typing the types as strings on parsing fixes this issue).
It seems there is currently no test for atom selection from LAMMPS DATA files,
so this might have gone undetected.

An error is raised for `ag.bonds` for an atom group that doesn't have bonds,
but this seems to also be the case for branch issue-363 right now, so I didn't
try to fix this.

An especially ugly place is where I tried to hack `_bondtypes`
into the existing code to remove duplicate bonds. This is also where
the code seems to be broken for non-explicit bond types. In this case,
I set `_bondtypes` to be a numpy array full of `None` with `dtype=object`.
It seems that `unique_rows` fails for this case, although it works fine
when `_bondtypes.dtype == "|S1"`.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by khuston** [comment]
I think I un-broke `atoms.bonds`, `atoms.angles`, etc. with this new commit

File: package/MDAnalysis/lib/util.py [hunk 1]
```diff
@@ -1164,32 +1164,3 @@ def wrapper(self, *args, **kwargs):
         return wrapper
 
     return cached_lookup
-
-
-def unique_rows(arr):
```
Comment (richardjgowers) [inline package/MDAnalysis/lib/util.py]: My functions aren't that bad :D

File: package/MDAnalysis/core/topologyattrs.py [hunk 2]
```diff
@@ -880,6 +881,9 @@ def __init__(self, values):
         Eg:  [(0, 1), (1, 2), (2, 3)]
         """
         self.values = values
+        if types is None:
```
Comment (richardjgowers) [inline package/MDAnalysis/core/topologyattrs.py]: I don't really like adding an array of Nones here, but I can see why you did it.  We'll fix this eventually

File: package/MDAnalysis/lib/util.py (hunk 1 above)
Comment (khuston) [inline package/MDAnalysis/lib/util.py]: Haha, PyCharm was telling me I deleted the only call made to it in the project (except in the testsuite). Should I un-delete it?

Comment (richardjgowers) [inline package/MDAnalysis/lib/util.py]: Ah right, I don't use a fancy IDE.   Errr keep it for now

**Comment by richardjgowers** [comment]
Looking good.  Is this everything you need to get the writer working?

**Comment by khuston** [comment]
Yes, this should be everything because whether a universe, timestep, or atom group is passed, the `bonds`, `angles`, `dihedrals`, and `impropers` will be accessible, along with their types.

To actually detect whether the types have been explicitly defined, I will have to go through protected member variable `_bondtypes` and verify they're not `None`.

Also, if the user passes an atomgroup that has an incomplete fragment, then the writer should either raise an error, or the resulting DATA file will be broken, because some bonds will include atoms that aren't written to the file.

**Comment by khuston** [comment]
I made another commit that fixes accessing empty TopologyGroups. Some parsers won't create the TopologyGroup at all, so `AttributeError` is raised on `atoms.impropers` if there are no impropers, but others (like TPR parser) create empty an TopologyGroup, and accessing it gave `IndexError: tuple index out of range` for me on branch issue-363. This error occurred in the section that removes duplicate bonds in `TopologyGroup.__init__()`, which I change here anyway, so I tried to fix this too.

**Comment by khuston** [comment]
Should I include the DATAWriter in this pull request? I have a working one now.

**Comment by richardjgowers** [comment]
New PR :)

**Pull request merged**